File: haskell/janken-couple/test/JankenCouple/HandSpec.hs
module JankenCouple.HandSpec where

import JankenCouple.Hand
import Test.Hspec
import Control.Monad

testCases :: [(Hand, Hand, Result)]
testCases = [ (Stone, Stone, Draw),
              (Stone, Scissors, Win),
              (Stone, Paper, Lose),
              (Scissors, Stone, Lose),
              (Scissors, Scissors, Draw),
              (Scissors, Paper, Win),
              (Paper, Stone, Win),
              (Paper, Scissors, Lose),
              (Paper, Paper, Draw) ]

spec :: Spec
spec = do
  describe "Hand" $ do
    forM_ testCases $ \(h1, h2, expected) ->
      describe ((show h1) ++ " vs") $ do
        describe (show h2) $ do
          it ("should be " ++ (show expected)) $ do
            (h1 `battleWith` h2) `shouldBe` expected

File: ruby/janken_multi/lib/hand_spec.rb

require 'rspec'
require_relative './hand'

describe 'JankenMulti::Hand' do
  describe 'Hand classes' do
    it 'Stone to_s should return "stone"' do
      expect(JankenMulti::Hand::STONE.to_s).to eq "stone"
    end

    it 'Paper to_s should return "paper"' do
      expect(JankenMulti::Hand::PAPER.to_s).to eq "paper"
    end

    it 'Scissors to_s should return "scissors"' do
      expect(JankenMulti::Hand::SCISSORS.to_s).to eq "scissors"
    end

    it 'Stone equality' do
      expect(JankenMulti::Hand::STONE).to eq JankenMulti::Hand::STONE
      expect(JankenMulti::Hand::STONE).not_to eq JankenMulti::Hand::PAPER
      expect(JankenMulti::Hand::STONE).not_to eq JankenMulti::Hand::SCISSORS
    end

    it 'Paper equality' do
      expect(JankenMulti::Hand::PAPER).to eq JankenMulti::Hand::PAPER
      expect(JankenMulti::Hand::PAPER).not_to eq JankenMulti::Hand::STONE
      expect(JankenMulti::Hand::PAPER).not_to eq JankenMulti::Hand::SCISSORS
    end

    it 'Scissors equality' do
      expect(JankenMulti::Hand::SCISSORS).to eq JankenMulti::Hand::SCISSORS
      expect(JankenMulti::Hand::SCISSORS).not_to eq JankenMulti::Hand::STONE
      expect(JankenMulti::Hand::SCISSORS).not_to eq JankenMulti::Hand::PAPER
    end
  end
end
